torchdynamo/mutation_guard.py

```diff
@@ -88,15 +78,24 @@ def is_dynamic_nn_module(obj):
     )
 
 
-def install_generation_tagging_new():
+def install_generation_tagging_init():
     """
-    Monkey patch torch.nn.Module.__new__ so we can detect nn.Module
-    instances created dynamically inside forward methods.
+    Monkey patch torch.nn.Module.__init__ and torch.nn.Module.__setstate__
+    so we can detect nn.Module instances created dynamically inside forward methods.
     """
-    assert (
-        Module.__dict__.get("__new__") is generation_tagging_new
-        or Module.__new__ is object.__new__
-    )
-    Module.__new__ = generation_tagging_new
-    GenerationTracker.db.clear()
+    init = Module.__init__
+
+    def patched_init(self, *args, **kwargs):
+        init(self, *args, **kwargs)
+        GenerationTracker.tag(self)
+
+    Module.__init__ = patched_init
```

Comment:
We should make sure this patch only happens once.  I think currently it will happen again and again (once for each call to torchdynamo.optimize)

Reply:
Yeah you're right, that's a good catch. I'll push a fix.